File: HoloLoom/vision/object_detector.py
"""
Object Detector - Real-time object detection for AR

Supports multiple backends:
    - TensorFlow.js COCO-SSD (web, 80 classes)
    - YOLOv8 (Python backend, 80 classes)
    - Custom models

Created: 2025-11-22
"""

from typing import List, Dict, Any, Optional
from dataclasses import dataclass
import logging
import numpy as np
from datetime import datetime

from HoloLoom.vision.protocol import (
    DetectedObject,
    BoundingBox,
    ObjectDetectorProtocol,
)


logger = logging.getLogger(__name__)


# ============================================================================
# COCO Class Labels (80 classes)
# ============================================================================

COCO_CLASSES = [
    "person", "bicycle", "car", "motorcycle", "airplane", "bus", "train",
    "truck", "boat", "traffic light", "fire hydrant", "stop sign",
    "parking meter", "bench", "bird", "cat", "dog", "horse", "sheep", "cow",
    "elephant", "bear", "zebra", "giraffe", "backpack", "umbrella", "handbag",
    "tie", "suitcase", "frisbee", "skis", "snowboard", "sports ball", "kite",
    "baseball bat", "baseball glove", "skateboard", "surfboard", "tennis racket",
    "bottle", "wine glass", "cup", "fork", "knife", "spoon", "bowl", "banana",
    "apple", "sandwich", "orange", "broccoli", "carrot", "hot dog", "pizza",
    "donut", "cake", "chair", "couch", "potted plant", "bed", "dining table",
    "toilet", "tv", "laptop", "mouse", "remote", "keyboard", "cell phone",
    "microwave", "oven", "toaster", "sink", "refrigerator", "book", "clock",
    "vase", "scissors", "teddy bear", "hair drier", "toothbrush"
]


# ============================================================================
# Object Detector Base
# ============================================================================

class ObjectDetector(ObjectDetectorProtocol):
    """
    Object detector with pluggable backends.

    Graceful degradation: Falls back to mock detector if backend unavailable.
    """

    def __init__(
        self,
        backend: str = "mock",
        confidence_threshold: float = 0.5,
        max_detections: int = 20,
        model_path: Optional[str] = None,
    ):
        """
        Initialize object detector.

        Args:
            backend: "tfjs", "yolo", "mock"
            confidence_threshold: Minimum confidence (0.0-1.0)
            max_detections: Maximum objects to return
            model_path: Path to custom model (optional)
        """
        self.backend = backend
        self.confidence_threshold = confidence_threshold
        self.max_detections = max_detections
        self.model_path = model_path
        self.model = None
        self.initialized = False

    async def initialize(self) -> bool:
        """Initialize detector backend"""
        try:
            if self.backend == "yolo":
                return await self._initialize_yolo()
            elif self.backend == "tfjs":
                # TensorFlow.js runs in browser, not here
                logger.info("TensorFlow.js detector runs in browser")
                self.initialized = True
                return True
            elif self.backend == "mock":
                logger.info("Using mock object detector")
                self.initialized = True
                return True
            else:
                logger.warning(f"Unknown backend: {self.backend}, using mock")
                self.backend = "mock"
                self.initialized = True
                return True

        except Exception as e:
            logger.error(f"Failed to initialize {self.backend}: {e}")
            logger.warning("Falling back to mock detector")
            self.backend = "mock"
            self.initialized = True
            return True  # Always succeed via fallback

    async def _initialize_yolo(self) -> bool:
        """Initialize YOLO backend (requires ultralytics)"""
        try:
            from ultralytics import YOLO

            # Load model
            model_path = self.model_path or "yolov8n.pt"  # Nano model (6.3MB)
            self.model = YOLO(model_path)
            self.initialized = True
            logger.info(f"YOLO model loaded: {model_path}")
            return True

        except ImportError:
            logger.warning("ultralytics not installed, install with: pip install ultralytics")
            return False
        except Exception as e:
            logger.error(f"Failed to load YOLO model: {e}")
            return False

    async def process_frame(self, frame: np.ndarray) -> List[DetectedObject]:
        """Process frame (delegates to detect_objects)"""
        return await self.detect_objects(frame, self.confidence_threshold)

    async def detect_objects(
        self,
        frame: np.ndarray,
        confidence_threshold: float = 0.5
    ) -> List[DetectedObject]:
        """
        Detect objects in frame.

        Args:
            frame: RGB image (HxWx3)
            confidence_threshold: Minimum confidence

        Returns:
            List of detected objects
        """
        if not self.initialized:
            await self.initialize()

        if self.backend == "yolo":
            return await self._detect_yolo(frame, confidence_threshold)
        elif self.backend == "mock":
            return self._detect_mock(frame, confidence_threshold)
        else:
            return []

    async def _detect_yolo(
        self,
        frame: np.ndarray,
        confidence_threshold: float
    ) -> List[DetectedObject]:
        """YOLO detection"""
        if not self.model:
            return []

        # Run inference
        results = self.model(frame, conf=confidence_threshold, verbose=False)

        # Parse results
        detections = []
        for result in results:
            boxes = result.boxes

            for i, box in enumerate(boxes):
                # Get bbox (xyxy format)
                xyxy = box.xyxy[0].cpu().numpy()
                x_min, y_min, x_max, y_max = xyxy

                # Normalize to 0-1
                h, w = frame.shape[:2]
                bbox = BoundingBox(
                    x_min=x_min / w,
                    y_min=y_min / h,
                    x_max=x_max / w,
                    y_max=y_max / h,
                )

                # Get class and confidence
                class_id = int(box.cls[0])
                confidence = float(box.conf[0])
                label = COCO_CLASSES[class_id] if class_id < len(COCO_CLASSES) else f"class_{class_id}"

                detection = DetectedObject(
                    id=f"obj_{i}_{datetime.now().timestamp()}",
                    label=label,
                    confidence=confidence,
                    bbox=bbox,
                    class_id=class_id,
                )

                detections.append(detection)

                # Limit detections
                if len(detections) >= self.max_detections:
                    break

        return detections

    def _detect_mock(
        self,
        frame: np.ndarray,
        confidence_threshold: float
    ) -> List[DetectedObject]:
        """Mock detection for testing"""
        # Return some dummy objects for demonstration
        return [
            DetectedObject(
                id="mock_obj_1",
                label="toolbox",
                confidence=0.85,
                bbox=BoundingBox(0.3, 0.4, 0.5, 0.7),
                class_id=0,
                metadata={"mock": True},
            ),
            DetectedObject(
                id="mock_obj_2",
                label="drill",
                confidence=0.92,
                bbox=BoundingBox(0.6, 0.3, 0.8, 0.6),
                class_id=1,
                metadata={"mock": True},
            ),
        ]

    async def cleanup(self) -> None:
        """Clean up resources"""
        self.model = None
        self.initialized = False

    def get_capabilities(self) -> Dict[str, bool]:
        """Get detector capabilities"""
        return {
            "object_detection": True,
            "real_time": self.backend in ["yolo", "tfjs"],
            "80_classes": self.backend in ["yolo", "tfjs"],
            "custom_models": self.backend == "yolo",
        }


# ============================================================================
# Factory
# ============================================================================

def create_object_detector(
    backend: str = "mock",
    **kwargs
) -> ObjectDetector:
    """
    Factory for creating object detectors.

    Args:
        backend: "yolo", "tfjs", "mock"
        **kwargs: Backend-specific options

    Returns:
        ObjectDetector instance
    """
    return ObjectDetector(backend=backend, **kwargs)


# ============================================================================
# Helper Functions
# ============================================================================

def filter_objects_by_type(
    objects: List[DetectedObject],
    labels: List[str]
) -> List[DetectedObject]:
    """Filter objects by label"""
    return [obj for obj in objects if obj.label in labels]


def get_largest_object(
    objects: List[DetectedObject]
) -> Optional[DetectedObject]:
    """Get object with largest bounding box"""
    if not objects:
        return None

    return max(objects, key=lambda obj: obj.bbox.area())


def get_closest_to_center(
    objects: List[DetectedObject],
    frame_center: Tuple[float, float] = (0.5, 0.5)
) -> Optional[DetectedObject]:
    """Get object closest to frame center"""
    if not objects:
        return None

    def distance_to_center(obj: DetectedObject) -> float:
        cx, cy = obj.bbox.center()
        return ((cx - frame_center[0])**2 + (cy - frame_center[1])**2)**0.5

    return min(objects, key=distance_to_center)


def non_max_suppression(
    objects: List[DetectedObject],
    iou_threshold: float = 0.5
) -> List[DetectedObject]:
    """
    Apply non-maximum suppression to remove overlapping boxes.

    Args:
        objects: List of detected objects
        iou_threshold: IoU threshold for suppression

    Returns:
        Filtered list of objects
    """
    if not objects:
        return []

    # Sort by confidence (descending)
    sorted_objs = sorted(objects, key=lambda x: x.confidence, reverse=True)

    keep = []
    while sorted_objs:
        # Keep highest confidence object
        current = sorted_objs.pop(0)
        keep.append(current)

        # Remove overlapping objects
        sorted_objs = [
            obj for obj in sorted_objs
            if current.bbox.iou(obj.bbox) < iou_threshold
        ]

    return keep

File: elle_unity/Assets/Scripts/Elle.Vision/ObjectDetector.cs
/**
 * ObjectDetector.cs - Real-time object detection for Unity
 *
 * Uses Unity Barracuda for on-device ML inference.
 * Supports YOLO, SSD, EfficientDet models.
 *
 * Created: 2025-11-24
 */

using System.Collections.Generic;
using UnityEngine;
using Unity.Barracuda;
using Elle.Core;

namespace Elle.Vision
{
    /// <summary>
    /// Real-time object detection using Unity Barracuda (on-device ML).
    /// Detects objects in camera feed and provides 3D positions.
    /// </summary>
    public class ObjectDetector : MonoBehaviour
    {
        [Header("Model Configuration")]
        [SerializeField] private NNModel modelAsset;
        [SerializeField] private TextAsset labelsAsset;
        [SerializeField] private WorkerFactory.Type workerType = WorkerFactory.Type.ComputePrecompiled;

        [Header("Detection Configuration")]
        [SerializeField] private Camera arCamera;
        [SerializeField] private int inputWidth = 640;
        [SerializeField] private int inputHeight = 480;
        [SerializeField] private float confidenceThreshold = 0.5f;
        [SerializeField] private int maxDetections = 20;
        [SerializeField] private float detectionInterval = 0.1f; // 10 FPS

        [Header("Debug")]
        [SerializeField] private bool visualizeDetections = true;
        [SerializeField] private bool logDetections = false;

        // Barracuda inference
        private IWorker worker;
        private Model model;
        private RenderTexture renderTexture;
        private Texture2D inputTexture;

        // Labels
        private string[] labels;

        // Current detections
        public List<DetectedObject> CurrentDetections { get; private set; } = new List<DetectedObject>();

        // State
        private float lastDetectionTime;
        private bool isProcessing = false;

        // Events
        public event System.Action<List<DetectedObject>> OnObjectsDetected;

        void Start()
        {
            if (arCamera == null)
            {
                arCamera = Camera.main;
            }

            InitializeModel();
            InitializeRenderTexture();
            LoadLabels();
        }

        /// <summary>
        /// Initialize Barracuda model
        /// </summary>
        void InitializeModel()
        {
            if (modelAsset == null)
            {
                Debug.LogWarning("No model asset assigned - object detection disabled");
                return;
            }

            try
            {
                model = ModelLoader.Load(modelAsset);
                worker = WorkerFactory.CreateWorker(workerType, model);
                Debug.Log($"✅ Object detection model loaded: {model.name}");
            }
            catch (System.Exception ex)
            {
                Debug.LogError($"Failed to load model: {ex.Message}");
            }
        }

        /// <summary>
        /// Initialize render texture for camera feed
        /// </summary>
        void InitializeRenderTexture()
        {
            if (arCamera == null)
                return;

            renderTexture = new RenderTexture(inputWidth, inputHeight, 24);
            arCamera.targetTexture = renderTexture;

            inputTexture = new Texture2D(inputWidth, inputHeight, TextureFormat.RGB24, false);

            Debug.Log($"Render texture initialized: {inputWidth}x{inputHeight}");
        }

        /// <summary>
        /// Load class labels from text file
        /// </summary>
        void LoadLabels()
        {
            if (labelsAsset == null)
            {
                Debug.LogWarning("No labels asset assigned - using default COCO labels");
                labels = GetDefaultCOCOLabels();
                return;
            }

            var lines = labelsAsset.text.Split('\n');
            labels = new string[lines.Length];

            for (int i = 0; i < lines.Length; i++)
            {
                labels[i] = lines[i].Trim();
            }

            Debug.Log($"Loaded {labels.Length} class labels");
        }

        void Update()
        {
            if (worker == null || isProcessing)
                return;

            // Throttle detection frequency
            if (Time.time - lastDetectionTime < detectionInterval)
                return;

            DetectObjectsAsync();
        }

        /// <summary>
        /// Run object detection on current camera frame
        /// </summary>
        async void DetectObjectsAsync()
        {
            isProcessing = true;
            lastDetectionTime = Time.time;

            try
            {
                // Capture camera frame
                CaptureFrame();

                // Convert to tensor
                var tensor = new Tensor(inputTexture, 3);

                // Run inference
                worker.Execute(tensor);
                var output = worker.PeekOutput();

                // Parse detections
                var detections = ParseDetections(output);

                // Update current detections
                CurrentDetections = detections;

                // Trigger event
                OnObjectsDetected?.Invoke(detections);

                if (logDetections && detections.Count > 0)
                {
                    Debug.Log($"Detected {detections.Count} objects");
                }

                // Cleanup
                tensor.Dispose();
            }
            catch (System.Exception ex)
            {
                Debug.LogError($"Detection error: {ex.Message}");
            }
            finally
            {
                isProcessing = false;
            }
        }

        /// <summary>
        /// Capture current camera frame to texture
        /// </summary>
        void CaptureFrame()
        {
            if (renderTexture == null || inputTexture == null)
                return;

            RenderTexture.active = renderTexture;
            inputTexture.ReadPixels(new Rect(0, 0, inputWidth, inputHeight), 0, 0);
            inputTexture.Apply();
            RenderTexture.active = null;
        }

        /// <summary>
        /// Parse detection output from model
        /// Note: Format depends on model type (YOLO, SSD, etc.)
        /// </summary>
        List<DetectedObject> ParseDetections(Tensor output)
        {
            var detections = new List<DetectedObject>();

            // This is a simplified parser - actual implementation depends on model
            // YOLO format: [batch, num_detections, 85] where 85 = [x, y, w, h, conf, 80 classes]
            // SSD format: [batch, num_detections, 7] where 7 = [?, class, conf, x1, y1, x2, y2]

            int numDetections = output.shape[1];
            int numClasses = labels.Length;

            for (int i = 0; i < numDetections && detections.Count < maxDetections; i++)
            {
                // Extract detection (format depends on model)
                float confidence = output[0, i, 4]; // Confidence score

                if (confidence < confidenceThreshold)
                    continue;

                // Get class with highest probability
                int classId = 0;
                float maxProb = 0f;

                for (int c = 0; c < numClasses; c++)
                {
                    float prob = output[0, i, 5 + c];
                    if (prob > maxProb)
                    {
                        maxProb = prob;
                        classId = c;
                    }
                }

                // Bounding box (normalized 0-1)
                float x = output[0, i, 0];
                float y = output[0, i, 1];
                float w = output[0, i, 2];
                float h = output[0, i, 3];

                // Convert to 3D position (simplified - assumes flat plane)
                var position = BoundingBoxToWorldPosition(x, y, w, h);

                var detection = new DetectedObject
                {
                    id = $"obj_{i}_{Time.frameCount}",
                    label = classId < labels.Length ? labels[classId] : $"class_{classId}",
                    confidence = confidence * maxProb,
                    position = position,
                    objectType = InferObjectType(classId < labels.Length ? labels[classId] : "")
                };

                detections.Add(detection);
            }

            return detections;
        }

        /// <summary>
        /// Convert 2D bounding box to 3D world position (simplified)
        /// </summary>
        Vector3 BoundingBoxToWorldPosition(float x, float y, float w, float h)
        {
            if (arCamera == null)
                return Vector3.zero;

            // Center of bounding box in normalized coords (0-1)
            var centerX = x;
            var centerY = y;

            // Convert to viewport point
            var viewportPoint = new Vector3(centerX, centerY, 1.5f); // Assume 1.5m depth

            // Convert to world position
            var worldPos = arCamera.ViewportToWorldPoint(viewportPoint);

            return worldPos;
        }

        /// <summary>
        /// Infer object type from label (for HoloLoom context)
        /// </summary>
        string InferObjectType(string label)
        {
            // Map labels to semantic types
            var typeMap = new Dictionary<string, string>
            {
                { "person", "person" },
                { "chair", "furniture" },
                { "couch", "furniture" },
                { "bed", "furniture" },
                { "dining table", "furniture" },
                { "laptop", "electronics" },
                { "cell phone", "electronics" },
                { "tv", "electronics" },
                { "bottle", "container" },
                { "cup", "container" },
                { "bowl", "container" },
                { "book", "document" },
                { "scissors", "tool" },
                { "knife", "tool" },
                { "spoon", "utensil" },
                { "fork", "utensil" }
            };

            return typeMap.ContainsKey(label) ? typeMap[label] : "object";
        }

        /// <summary>
        /// Get default COCO dataset labels (80 classes)
        /// </summary>
        string[] GetDefaultCOCOLabels()
        {
            return new string[]
            {
                "person", "bicycle", "car", "motorcycle", "airplane",
                "bus", "train", "truck", "boat", "traffic light",
                "fire hydrant", "stop sign", "parking meter", "bench", "bird",
                "cat", "dog", "horse", "sheep", "cow",
                "elephant", "bear", "zebra", "giraffe", "backpack",
                "umbrella", "handbag", "tie", "suitcase", "frisbee",
                "skis", "snowboard", "sports ball", "kite", "baseball bat",
                "baseball glove", "skateboard", "surfboard", "tennis racket", "bottle",
                "wine glass", "cup", "fork", "knife", "spoon",
                "bowl", "banana", "apple", "sandwich", "orange",
                "broccoli", "carrot", "hot dog", "pizza", "donut",
                "cake", "chair", "couch", "potted plant", "bed",
                "dining table", "toilet", "tv", "laptop", "mouse",
                "remote", "keyboard", "cell phone", "microwave", "oven",
                "toaster", "sink", "refrigerator", "book", "clock",
                "vase", "scissors", "teddy bear", "hair drier", "toothbrush"
            };
        }

        /// <summary>
        /// Visualize detections (for debugging)
        /// </summary>
        void OnGUI()
        {
            if (!visualizeDetections || CurrentDetections.Count == 0)
                return;

            foreach (var detection in CurrentDetections)
            {
                var screenPos = arCamera.WorldToScreenPoint(detection.position);

                if (screenPos.z > 0)
                {
                    var label = $"{detection.label} ({detection.confidence:F2})";
                    var labelSize = GUI.skin.label.CalcSize(new GUIContent(label));

                    // Draw background box
                    var rect = new Rect(
                        screenPos.x - labelSize.x / 2,
                        Screen.height - screenPos.y - labelSize.y / 2,
                        labelSize.x + 10,
                        labelSize.y + 5
                    );

                    GUI.Box(rect, "");

                    // Draw label
                    GUI.Label(rect, label);
                }
            }
        }

        void OnDestroy()
        {
            worker?.Dispose();

            if (renderTexture != null)
            {
                arCamera.targetTexture = null;
                Destroy(renderTexture);
            }

            if (inputTexture != null)
            {
                Destroy(inputTexture);
            }
        }
    }
}
